File: Java/3397. Maximum Number of Distinct Elements After Operations.java
class Solution {
    public int maxDistinctElements(int[] nums, int k) {
        Arrays.sort(nums);
        int ans = 0, pre = Integer.MIN_VALUE;

        for (int n: nums) {
            int cur = Math.min(Math.max(pre + 1, n - k), n + k);
            if (cur > pre) {
                ans++;
                pre = cur;
            }
        }

        return ans;
    }
}

File: Python3/3397. Maximum Number of Distinct Elements After Operations.py
class Solution:
    def maxDistinctElements(self, nums: List[int], k: int) -> int:
        nums.sort()
        pre = nums[0] - k
        ans = 1

        for n in nums[1:]:
            cur = min(max(pre + 1, n - k), n + k)
            if cur > pre:
                ans += 1
                pre = cur

        return ans